### ISSUE TITLE
spring security reads only a single configuration

### ISSUE THREAD

**Issue by Captain-P-Goldfish** [spring-projects/spring-security#5745]
Hi I have tried a lot of things now and I start to become frustrated about it... I am trying to add two different configurations in my application for spring security. Both have to secure different parts of the application. But the problem is, that spring always reads only the first configuration in load order. Any other configurations are ignored completely. The configure methods of these classes are called but the result gets lost... My code looks like follows

    
    /** */
     @Configuration
      @Order(LOAD_ORDER + 1)
      public static class ClientAuthenticationConfiguration extends WebSecurityConfigurerAdapter
      {
        /**
         * the server port for the configuration endpoints
         */
        @Getter
        @Value("${server.port}")
        protected int configurationPort;
    
        /**
         * this configuration will catch the client-certificate contents that have been provided during x509
         * client authentication and will read the common-name from the distinguished name of the certificate that
         * is then passed by spring to the {@link #userDetailsService()}
         */
        @Override
        protected void configure(HttpSecurity http) throws Exception
        {
          http.authorizeRequests()
              .antMatchers(ContextPaths.CLIENT_CONFIGURATION + "/**",
                           ContextPaths.SERVICE_PROVIDER_CONFIGURATION + "/**",
                           ContextPaths.APPLICATION_CONFIGURATION + "/**")
              .authenticated()
              .and()
              .x509()
              .subjectPrincipalRegex("CN=(.*?)(?:,|$)")
              .userDetailsService(userDetailsService());
        }
    
        /**
         * @return the user details service that will provide an authenticated user with a role based on the
         *         provided x509 certificate
         */
        @Bean
        public UserDetailsService userDetailsService()
        {
          return username -> {
            if (username != null)
            {
              return new User(username, "", new ArrayList<>());
            }
            return null;
          };
        }
      }
    
      @Configuration
      @Order(LOAD_ORDER + 2)
      public static class OAuth2SecurityConfiguration extends WebSecurityConfigurerAdapter
      {
    
        /**
         * the server port for the configuration endpoints
         */
        @Getter
        @Value("${server.port}")
        protected int configurationPort;
    
        /**
         * {@inheritDoc}
         */
        @Override
        protected void configure(HttpSecurity http) throws Exception
        {
          http.csrf()
              .disable()
              .authorizeRequests()
              .antMatchers(ContextPaths.LOGIN,
                           ContextPaths.AUTHORIZATION_ENDPOINT,
                           ContextPaths.TOKEN_ENDPOINT,
                           ContextPaths.USER_INFO_ENDPOINT,
                           SamlRedirectContextPaths.SAML_CONSUMER_URL,
                           SamlRedirectContextPaths.USER_REDIRECT_ENDPOINT)
              .permitAll()
              .and()
              .formLogin()
              .loginPage(ContextPaths.LOGIN + "?npa");
    
          http.addFilterBefore(new ConfigurationEndpointsSecurityFilter(configurationPort), LogoutFilter.class);
        }
    
        /**
         * this bean is necessary to declare for spring-security-oauth2 to validate the client_secrets in the
         * token request
         */
        @Bean
        public PasswordEncoder getBCryptPasswordEncoder()
        {
          return new BCryptPasswordEncoder()
          {
    
            @Override
            public boolean matches(CharSequence rawPassword, String encodedPassword)
            {
              return super.matches(rawPassword, encodedPassword)
                     || StringUtils.equals(rawPassword, encodedPassword);
            }
          };
        }
      }

only the first configuration is used. Not even the filter from the second configuration is getting registered in the filter-chain eventhough the method is executed

**Comment by rwinch** [comment]
Only the first configuration that matches will be used. For example, you should specify something like

```
http
    .mvcMatchers("/admin/**")
    ...
```

Keep in mind that the matcher on HttpSecurity needs to be invoked for multiple HttpSecurity instances. The matchers on authorizeRequests are only for authorization rules and do not impact the entire HttpSecurity instance like the matchers on HttpSecurity do. See https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#multiple-httpsecurity

**Comment by rwinch** [comment]
I'm closing this as answered as I believe it has been answered. Please reopen with additional details if it does not solve your problem.

**Comment by Captain-P-Goldfish** [comment]
how would you explain then that the filter is not registered in the second subclass?
at least this should work doesn't it?

**Comment by rwinch** [comment]
It is likely getting registered but it won't be seen unless you reach the proper configuration. They are totally isolated from one another. In any event, it does not make much sense to troubleshoot what is happening as the configuration above will not work for the reasons I already described.

If you continue to have issues after resolving the issue I pointed out put together a complete sample of your issue and clearly describe what you expect to happen and what is not happening.
